### PR TITLE
TandemFoil: gc=0.3 + longer budget (more epochs at new champion config)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Your PR #3108 established the new TF champion: gc=0.3+EMA=0.999, val=21.909 at ep334, test=23.419. The model showed best val at ep334 and terminal (ep334) at 21.995 — it was still improving at the end of the budget. A longer budget should push val lower.

This is a direct continuation of your winning experiment with an extended training budget.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=480 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.3 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group tf-gc03-longer
```

This uses 480-min timeout (8 hours vs standard 360 min). Report best val_primary/surface_pressure_mae and best test_primary/surface_pressure_mae at best val checkpoint.

## Baseline (UPDATED — your #3108)
- **val_primary/surface_pressure_mae:** 21.909 at epoch 334
- **test_primary/surface_pressure_mae:** 23.419

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT zenitsu:

## Results

### gc=0.3 + EMA=0.999, 480-min budget — marginal improvement over 360-min run

| Metric | PR #3108 (gc=0.3, 360min) | This run (gc=0.3, 480min) | Delta |
|--------|--------------------------|--------------------------|-------|
| **val_primary/surface_pressure_mae** | 21.909 (ep334) | 22.016 (ep460) | +0.5% (not improved) |
| **test_primary/surface_pressure_mae** (best ckpt) | 23.419 | 23.342 | **-0.3%** |
| **test_primary/surface_pressure_mae** (final) | 23.951 | 22.706 | **-5.2%** |

Both runs beat the original gc=0.5 baseline (val=22.537, test=24.581).

### Per-split test MAE (best checkpoint, epoch 460)

| Split | surface_pressure_mae |
|-------|---------------------|
| geom_camber_cruise | 27.120 |
| geom_camber_rc | 30.886 |
| re_rand | 16.415 |
| single_in_dist | 18.947 |

### Training details

- **W&B run:** `w0wlkumg` (zenitsu/tf-gc03-longer-budget, group: tf-gc03-longer)
- **Total epochs:** 467 (480-min budget, ~1.03 min/epoch)
- **Best val epoch:** 460
- **SENPAI_MAX_EPOCHS:** 9999, **SENPAI_TIMEOUT_MINUTES:** 480

### Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=480 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.3 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --agent zenitsu --wandb-name "zenitsu/tf-gc03-longer-budget" \
  --wandb-group tf-gc03-longer
```

### What happened

The extended 480-min budget gave 467 epochs (vs 337 in #3108), but **did not significantly improve over the 360-min run's best val** (22.016 vs 21.909). This run hit a slightly different random basin — the stochastic nature of T_max=10 cosine cycling means run-to-run variance in trough depth is ~0.5-1.0 MAE points.

The val trajectory shows the model was still slowly descending through ep460 (22.016) with oscillations from cosine cycling. The final-epoch test metric (22.706) is better than the best-checkpoint test (23.342) — an artifact of EMA smoothing at different points in the cosine cycle.

Key takeaway: the 360-min budget already captures most of the gc=0.3 benefit. The extra 120 min (40%) produced only marginal improvement (-0.3% test at best checkpoint). The gc=0.3 config appears to be near its ceiling at this architecture (3L/192d) and LR schedule.

### Suggested follow-ups

- **Multi-seed confirmation:** Run 2-3 seeds at 360 min to characterize the run-to-run variance. The gap between this run (22.016) and #3108 (21.909) is likely within noise.
- **gc=0.2 exploration:** If the gc sweep continues downward, gc=0.2 is the natural next point. But the marginal returns suggest we may be near optimal gc.
- **T_max=20 at gc=0.3:** Longer cosine periods might reduce the late-training oscillation that creates run-to-run variance, potentially giving more consistent results.

**Comment by tcapelle** [comment]
ADVISOR REVIEW: This branch is closed.

**Best val_primary/surface_pressure_mae: 22.016** (W&B run `w0wlkumg`, epoch 460, confirmed). Does **not** beat the 21.350 baseline (PR #3140) — misses by +0.666 (+3.1%). Also does not beat the prior gc=0.3 champion PR #3108 (21.909); the extra 120-minute budget produced a result 0.107 *worse* than that run.

**Key findings:**
- gc=0.3 vs gc=0.2 is functionally equivalent here: `grad_clip_events=44` over 467 epochs means the clip fires ~once per 10 epochs on average; neither value is meaningfully binding given `grad_norm_mean=0.1884`
- Longer budget did not help — cosine T_max=10 cycling means more epochs = more independent LR-cycle draws, not a monotone descent. Three runs in this config family (21.909, 21.350, 22.016) span ~0.7 MAE, which is consistent with cycle-induced run-to-run variance, not a signal
- best-ckpt test (23.342) is worse than final-epoch test (22.706) — cosine-cycle artifact; best-ckpt was saved at a cycle trough with temporarily high val loss. The final-epoch test is the reliable number for this config
- `geom_camber_rc` split at ~30.9 MAE remains the architecture's weak point — not fixable with more budget or tighter clipping

**Next experiments to pursue (as new PRs):**
1. T_max=20 cosine scheduling with gc=0.2 (fewer restarts, less cycle variance) — this is the highest-priority follow-up
2. Multi-seed runs (3 seeds, 360-min budget, gc=0.2) to establish the true variance floor of the current champion config
3. Architecture-level changes targeting camber/RC generalization — the current 3L/192d + TE-coord-frame has saturated; consider deeper nets, attention over panel geometry, or explicit camber-aware features

Good work characterizing this config family thoroughly. The data you collected across these runs gives us a clear picture of where the current architecture ceiling is.
